### PR TITLE
Share code for submission running (extension/sandbox) and team submission info (extension/web)

### PR DESCRIPTION
**Pull request by dpoeschl**
**Added two typescript projects to `shared`**, in both cases deduplicating code between the referencing projects:
1) `submissionRunner`
    - Used by the `sandbox` and `extension` to run team code, exports runner methods (`runJava` etc.) and related types

2) `extensionWeb`
    - So far only exports types for communicating contest state between `web` and `extension` components, but plan to add more if we accept this restructuring

**Dev Process**
- Building main components automatically picks up latest changes to the `/shared/*` projects, even if they haven't been built yet
- `npm run watch` for the `extension` notices file saves in the `/shared/*` projects as expected

**Docker**
- The docker filesystems are restructured a bit. For example, instead of the `web` project living at `/app`, it now lives at `/app/web` with shared projects under `/app/shared/...`, mimicking how they relate in the project & allowing builds to work in that context.
- ❓Not sure if the docker part is fully correct here. They build and start running, but in both cases don't see environment variables. Could be a problem with my current setup, just haven't had the chance to verify, so that's why this PR is in draft.

**Misc.**
To avoid rationalizing the existing module kinds / build setups / etc. between the components, using `.cts` files in the shared projects was the only way I found to make them consumable everywhere.

### CODE REVIEW

File: shared/submissionRunner/types.cts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	input: string;
+	outputCallback?: (data: string) => void;
+}
+
+export type IRunnerReturn =
+	{ success: true; killFunc: () => void; runResult: Promise<RunResult> } |
+	{ success: false; runResult: RunResult };
+
+export type IRunner<T extends IRunnerParams = IRunnerParams> = 
+	(params: T) => Promise<IRunnerReturn>;
```
Comment (dpoeschl) [inline shared/submissionRunner/types.cts:36]: It's a bit lost in this PR, but I had to change these types. 

The `IRunner<...>` used to be a method that returned `IRunnerReturn`, which was a type expressed as `Promise< {success case} | {failure case} >`.

That structure threw errors in this restructured shared project, so I changed it to have the `IRunner<...>` method to return a `Promise<IReturnRunner>` (which makes sense, it's a method you should await and you can see that more clearly on the signature now), with `IRunnerReturn` just directly `{success case} | {failure case}` (not wrapped in a `Promise`, which also makes sense).

The old structure must've been fine in one `tsconfig` setup, but this one works here and also is easier to read IMO.

(The change of `IRunner<...>` from an interface to type was personal preference, I find the interface-with-one-unnamed-method pattern weird, and some tools recommend switching to this as well. Let me know if you really prefer it the previous way.)

Comment (orosmatthew) [reply shared/submissionRunner/types.cts:36]: Seems good to be, I agree the previous types were pretty ugly

**Comment by orosmatthew** [comment]
I seem to have fixed the environment variables issue by using an `entry.sh` script instead of `npm run start` for the sandbox. Also, the `entry.sh` files must use Unix LF separators or else docker will say the file doesn't exist (such nasty behavior). I gave everything a quick run through as it is after my two commits I added and things seemed to work fine. Let me know if you are happy and then I can test out the merge into master before merging the PR

**Comment by orosmatthew** [comment]
I also just fixed some imports on the web side. I used the correct alias for importing the shared code which fixed errors reported by `npm run check` (SvelteKit Error checker) and also remove the path reference in the `tsconfig` file for web as you only need to create aliases in the svelte config file only

**Pull request merged**